Bugfix for GL waveforms with QT 5.14 on Windows.

JosepMaJAZ opened the pull request:
This is a workaround for a bug that has appeared on Windows with GL waveforms when updating to QT 5.14.2.

For some unclear reason, when we first create the GL waveforms, its window thinks it is not exposed (exposed is not the same as visible https://doc.qt.io/qt-5/qwindow.html#exposeEvent ) and so the rendering doesn't take place.

If we start with software rendering and switch to GL, then that first change didn't work either, and that's why I had to modify the dlgPrefWaveform too.

Not sure if it is the best solution, or also if there is the possibility to force this with less changes. I am open to opinions

JosepMaJAZ committed:
Bugfix for GL waveforms with QT 5.14 on Windows.

Be-ing commented:
This seems like an acceptable hack to me. I tested on Linux and did not notice regressions with any waveform renderer. I would like to hear @daschuer's thoughts before merging.

Holzhaus commented on src/mixxx.cpp:
```diff
@@ -449,6 +449,11 @@ void MixxxMainWindow::initialize(QApplication* pApp, const CmdlineArgs& args) {
+    connect(this,
+            &MixxxMainWindow::newSkinLoaded,
+            WaveformWidgetFactory::instance(),
+            &WaveformWidgetFactory::onSkinLoadFinished);
```
We currently have two different naming schemes in our codebase: `onSomeSignal` and `slotSomeSignal`. We should agree on a single way of naming things. I think the latter is more prevalent atm but the former is nicer on the eyes IMHO. Opinions?

Be-ing replied:
I don't have a strong opinion either way, but it's easier to go with `slotSomeSignal` because it is already used more.

JosepMaJAZ replied:
As I say below, the newSkinLoaded signal already existed and I duplicated the connect that existed for Library, even the naming. 
But indeed, making it explicit that this is expected to be called from a signal is better.
So, what do I do?

Holzhaus commented on src/mixxx.cpp:
```diff
@@ -449,6 +449,11 @@ void MixxxMainWindow::initialize(QApplication* pApp, const CmdlineArgs& args) {
             &Library::onSkinLoadFinished);
 
+    connect(this,
+            &MixxxMainWindow::newSkinLoaded,
```
How about `skinLoaded`? And I think the slot should also be named `onSkinLoaded` or `slotSkinLoaded` respectively.

JosepMaJAZ replied:
The name of the signal already existed (it is connected on the previous sentence).
Should I still change it?

Holzhaus commented on src/preferences/dialog/dlgprefwaveform.cpp:
```diff
@@ -196,7 +196,12 @@ void DlgPrefWaveform::slotSetWaveformType(int index) {
     if (index < 0) {
         return;
     }
+    // This regenerates twice the waveforms because of a bug found on Windows
+    // where the first one fails.
+    // The problem is that the window of the widget thinks that it is not exposed.
+    // (https://doc.qt.io/qt-5/qwindow.html#exposeEvent )
```
```suggestion
    // This regenerates the waveforms twice because of a bug found on Windows
    // where the first one fails.
    // The problem is that the window of the widget thinks that it is not exposed.
    // (https://doc.qt.io/qt-5/qwindow.html#exposeEvent )
    // TODO: Remove this when it has been fixed upstream.
```

Holzhaus replied:
Maybe add a link to a Qt bug report?

Be-ing replied:
I looked briefly but didn't find one.

Holzhaus commented on src/waveform/waveformwidgetfactory.cpp:
```diff
@@ -359,6 +364,16 @@ void WaveformWidgetFactory::addTimerListener(QWidget* pWidget) {
 
+
+void WaveformWidgetFactory::onSkinLoadFinished() {
+    // This regenerates twice the waveforms because of a bug found on Windows
+    // where the first one fails.
+    // The problem is that the window of the widget thinks that it is not exposed.
+    // (https://doc.qt.io/qt-5/qwindow.html#exposeEvent )
```
```suggestion
    // This regenerates the waveforms twice because of a bug found on Windows
    // where the first one fails.
    // The problem is that the window of the widget thinks that it is not exposed.
    // (https://doc.qt.io/qt-5/qwindow.html#exposeEvent )
    // TODO: Remove this when it has been fixed upstream.
```

Holzhaus commented on src/waveform/waveformwidgetfactory.cpp:
```diff
@@ -430,31 +447,40 @@ int WaveformWidgetFactory::getVSyncType() {
+        desired = empty;
+    }
+    return setWidgetTypeFromHandle(desired, true);
+}
+bool WaveformWidgetFactory::setWidgetTypeFromHandle(int handleIndex, bool force) {
```
```suggestion
}

bool WaveformWidgetFactory::setWidgetTypeFromHandle(int handleIndex, bool force) {
```

Holzhaus commented on src/waveform/waveformwidgetfactory.cpp:
```diff
@@ -973,3 +999,21 @@ void WaveformWidgetFactory::startVSync(GuiTick* pGuiTick, VisualsManager* pVisua
+        type = m_waveformWidgetHandles[index].m_type;
+    }
+    return type;
+}
+int WaveformWidgetFactory::findHandleIndexFromType(WaveformWidgetType::Type type) {
```
```suggestion
}

int WaveformWidgetFactory::findHandleIndexFromType(WaveformWidgetType::Type type) {
```

Holzhaus commented on src/waveform/waveformwidgetfactory.h:
```diff
@@ -64,8 +64,9 @@ class WaveformWidgetFactory : public QObject, public Singleton<WaveformWidgetFac
 
-    //creates the waveform widget and bind it to the viewer
-    //clean-up every thing if needed
+    // creates the waveform widget using the type set with setWidgetType
+    // and binds it to the viewer.
+    // deletes older widget and resets positions to config defaults
```
I propose that we use 3 slashes for comments that document whole functions, so that they are picked up by doxygen.  Please change this for the other occurrences, too.

```suggestion
    /// Creates the waveform widget using the type set with setWidgetType
    /// and binds it to the viewer.
    /// Deletes older widget and resets positions to config defaults
```

Holzhaus reviewed (changes requested):
Thanks. This looks good, I just have a few comments regarding code documentation and function naming.

JosepMaJAZ committed:
Update src/preferences/dialog/dlgprefwaveform.cpp

Better comments

Co-Authored-By: Jan Holthuis <holthuis.jan@googlemail.com>

JosepMaJAZ committed:
Fixes from the PR (naming , doxygen and spaces)

JosepMaJAZ committed:
Fixes from the PR (naming , doxygen and spaces)

JosepMaJAZ committed:
Fixes from the PR (naming , doxygen and spaces)

Be-ing commented:
Please include some context in your commit messages. "Fixes from the PR" says very little. "MixxxMainWindow: rename newSkinLoaded signal to skinLoaded" is a much more useful message when looking back at the git history years later.

JosepMaJAZ committed:
Restored autodetection of the build environment to use on appveyor.yml. (Was removed when changed to CMake). Also added autodetection of QT version.
Added autodetection of QT version on build_mixxx.bat ( to build with scons. No longer used on appveyor)

JosepMaJAZ committed:
Merge branch 'master' into GL_Qt514_windows-fix

Holzhaus commented on appveyor.yml:
```diff
@@ -149,15 +149,13 @@ for:
-    - set ENVIRONMENT_NAME=%%ENVIRONMENT_NAME:CONFIGURATION=%configuration%%%
+    - set /P ENVIRONMENT_NAME=<build/windows/golden_environment
+    - call set ENVIRONMENT_NAME=%%ENVIRONMENT_NAME:PLATFORM=%platform%%%
+    - call set ENVIRONMENT_NAME=%%ENVIRONMENT_NAME:CONFIGURATION=%configuration%%%
```
What does `call` do here?

JosepMaJAZ replied:
Possibly it is unneeded, but that's exactly how it was before the CMake commit that removed it.

Holzhaus replied:
Can you check if it works without that? If so, please remove it.

Holzhaus commented on appveyor.yml:
```diff
@@ -172,8 +170,11 @@ for:
     - python -m pip install git+https://github.com/frerich/clcache.git
 
   before_build:
+    - FOR /D %%G IN (%WINLIB_PATH%\Qt-*) DO SET QT_PATH=%%G
```
...and some people say the unix command line is complicated.

JosepMaJAZ replied:
Hahahah! Yes , for loops in windows shell are a bit complicated. More so because depending on the parameter ( /D in this case ) it does different things. 
This concrete line can be understood as: for each directory in winlibpath that starts with Qt-, set QT_PATH to that absolute path 
( https://ss64.com/nt/for_d.html )

Holzhaus commented on src/waveform/waveformwidgetfactory.h:
```diff
@@ -64,8 +64,9 @@ class WaveformWidgetFactory : public QObject, public Singleton<WaveformWidgetFac
 
-    //creates the waveform widget and bind it to the viewer
-    //clean-up every thing if needed
+    /// creates the waveform widget using the type set with setWidgetType
+    /// and binds it to the viewer.
+    /// deletes older widget and resets positions to config defaults
```
```suggestion
    /// Creates the waveform widget using the type set with setWidgetType
    /// and binds it to the viewer.
    /// Deletes older widget and resets positions to config defaults.
```

Holzhaus commented on src/waveform/waveformwidgetfactory.h:
```diff
@@ -83,10 +84,20 @@ class WaveformWidgetFactory : public QObject, public Singleton<WaveformWidgetFac
+    /// Changes the widged type to that loaded from config and recreates them.
+    /// Used as a workaround on Windows due to a problem with GL and QT 5.14.2
+    bool setWidgetTypeFromConfig();
+    /// Changes the widget type and recreates them. Used from the pref. dialog
```
```suggestion
    /// Changes the widget type and recreates them. Used from the preferences
    /// dialog.
```

Holzhaus reviewed (changes requested):
I can't really comment on most of the changes and don't know if it fixes the issue on Windows since I'm on Linux, but I left some minor comments.

JosepMaJAZ committed:
Syntax from a bash script needs two %, but from commandline it needs just one.

JosepMaJAZ committed:
Improving comments

Co-Authored-By: Jan Holthuis <holthuis.jan@googlemail.com>

JosepMaJAZ committed:
Improving comments

Co-Authored-By: Jan Holthuis <holthuis.jan@googlemail.com>

JosepMaJAZ committed:
Forgot to allow mkdir to fail (cache restored but missing qt version).

JosepMaJAZ committed:
So it does need the "script" syntax, not the commandline syntax.

Be-ing commented:
CI is working except for pre-commit:
```
 pre-commit run --origin HEAD --source origin/HEAD --show-diff-on-failure

Check for byte-order marker..............................................Passed

Check for case conflicts.................................................Passed

Check JSON...........................................(no files to check)Skipped

Check for merge conflicts................................................Passed

Check Xml............................................(no files to check)Skipped

Check Yaml...............................................................Passed

Fix End of Files.........................................................Passed

Mixed line ending........................................................Passed

Trim Trailing Whitespace.................................................Failed

- hook id: trailing-whitespace

- exit code: 1

- files were modified by this hook

Fixing build/appveyor/build_mixxx.bat

Don't commit to branch...................................................Passed

eslint...............................................(no files to check)Skipped

clang-format.............................................................Passed

black................................................(no files to check)Skipped

flake8...............................................(no files to check)Skipped

qsscheck.............................................(no files to check)Skipped

line-length..............................................................Passed

All changes made by hooks:

diff --git a/build/appveyor/build_mixxx.bat b/build/appveyor/build_mixxx.bat

index f26c936ec..9507d2a5b 100644

--- a/build/appveyor/build_mixxx.bat

+++ b/build/appveyor/build_mixxx.bat

@@ -86,7 +86,7 @@ if "%4" == "skipinstaller" (

     if "%5" == "skipinstaller" (

        set PARAM_INSTALLER=

     ) else (

-       set PARAM_INSTALLER=makerelease 

+       set PARAM_INSTALLER=makerelease

     )

 )

 set WINLIB_DIR=%3

The command "pre-commit run --origin HEAD --source origin/HEAD --show-diff-on-failure" exited with 1.
```

Holzhaus commented:
OK, that means is there is trailing whitespace in that line.

Be-ing commented:
AppVeyor build succeeded, but now there is a merge conflict in  build/appveyor/build_mixxx.bat

JosepMaJAZ committed:
Merge remote-tracking branch 'remotes/upstream/master' into GL_Qt514_windows-fix

Be-ing commented:
Now there's another merge conflict.

daschuer commented on src/preferences/dialog/dlgprefwaveform.cpp:
```diff
@@ -196,7 +196,13 @@ void DlgPrefWaveform::slotSetWaveformType(int index) {
+    // (https://doc.qt.io/qt-5/qwindow.html#exposeEvent )
+    // TODO: Remove this when it has been fixed upstream.
     WaveformWidgetFactory::instance()->setWidgetTypeFromHandle(index);
+    WaveformWidgetFactory::instance()->setWidgetTypeFromHandle(index, true);
```
I think we  should guard it with the QT version and the effected OS. 
This way we will find this spot easily when we consider the next QT update.

JosepMaJAZ replied:
No. GL needs to be started at least once.
As I said, I also was reproducing the problem if starting with nonGl and then switching to a GL one from the preferences.
Ok, I can add if QT >= 5.14. Is that ok?

daschuer replied:
Yes, thank you.

daschuer commented on src/waveform/waveformwidgetfactory.cpp:
```diff
@@ -430,31 +447,42 @@ int WaveformWidgetFactory::getVSyncType() {
+        // type is acceptable
+        currentType = type;
+        if (m_config) {
+            m_config->set(ConfigKey("[Waveform]", "WaveformType"), ConfigValue((int)(currentType)));
```
Not your fault, but we prefer static_cast<int>()
You can also the setValue template.

The same below.

daschuer commented on src/waveform/waveformwidgetfactory.cpp:
```diff
@@ -430,31 +447,42 @@ int WaveformWidgetFactory::getVSyncType() {
+}
+
+bool WaveformWidgetFactory::setWidgetTypeFromHandle(int handleIndex, bool force) {
     if (handleIndex < 0 || handleIndex >= (int)m_waveformWidgetHandles.size()) {
```
size() returns already int. 

The same below.

JosepMaJAZ replied:
Oh... I suppose I had size_t in mind probably...

daschuer commented on src/waveform/waveformwidgetfactory.cpp:
```diff
@@ -973,3 +1001,22 @@ void WaveformWidgetFactory::startVSync(GuiTick* pGuiTick, VisualsManager* pVisua
+    if (index >= 0 && index < (int)m_waveformWidgetHandles.size()) {
+        type = m_waveformWidgetHandles[index].m_type;
+    }
+    return type;
```
So this returns an invalid type. Is that case handled in the calling code?

daschuer replied:
No, because there is no caller. Remove?

JosepMaJAZ replied:
I added it because it is the opposite of the next method, and I was going to use it first on setWidgetTypeFromHandle, but I found that I couldn't express the same if I used the function.

The logic of returning an invalid type instead of the empty type is so that one can detect that it has called the method with an invalid index.

daschuer commented on src/waveform/waveformwidgetfactory.cpp:
```diff
@@ -359,6 +364,16 @@ void WaveformWidgetFactory::addTimerListener(QWidget* pWidget) {
+    // where the first one fails.
+    // The problem is that the window of the widget thinks that it is not exposed.
+    // (https://doc.qt.io/qt-5/qwindow.html#exposeEvent )
+    setWidgetTypeFromConfig();
```
Please guard this as well. 
Here in both cases the force flag is set in the later call. 
Is this intended?

JosepMaJAZ replied:
The force parameter was added simply so that it would not cause a no-op if the same waveform is already selected. It's just a "I don't know what's setup, but probably I need to force a replace".

daschuer replied:
Ah, ok do it doesn't matter.
Can we always force than?

daschuer commented on src/waveform/waveformwidgetfactory.cpp:
```diff
@@ -463,7 +491,7 @@ bool WaveformWidgetFactory::setWidgetTypeFromHandle(int handleIndex) {
 
     WaveformWidgetAbstractHandle& handle = m_waveformWidgetHandles[handleIndex];
-    if (handle.m_type == m_type) {
+    if (handle.m_type == m_type && !force) {
```
I wonder what code exactly in the key for the fix? 

Is it: 
holder.m_waveformWidget = widget;
viewer->setWaveformWidget(widget);
widget->getWidget()->show();
viewer->update();

If we would know that, we can call this in the beginning of the loop and the patch becomes real small.

Can you check it?

JosepMaJAZ replied:
Remember that RJ also made some attempts in  #2681 that didn't work here.

daschuer replied:
Ok, so no need to do this again.

JosepMaJAZ replied:
I've been trying for a while adding the waveformviewer bits of that PR to mine (and commenting my doublecreation) but nothing worked.
I still don't quite get the logic, but with the additional logging on RJ's PR, I see how the exposed flag changes:

When I create the object once (which happens way before these log lines)
```
Debug [Main]: WWaveformViewer(0x2e8c53cec80) resizeEvent
Debug [Main]: WWaveformViewer(0x2e8c53cec80) isVisible() true QSize(1507, 73)
Debug [Main]: WWaveformViewer(0x2e8c53cec80) isVisible() WWidgetGroup(0x2e8ad34ad80, name="Waveform") QSize(1585, 73)
Debug [Main]: WWaveformViewer(0x2e8c53cec80) widget isVisible() true QSize(22, 10)
Debug [Main]: WWaveformViewer(0x2e8c53cec80) widget isVisible() true QSize(1507, 73)
Debug [Main]: QWidgetWindow(0x2e8d0246070, name="QGLWidgetClassWindow") vis true QSize(1507, 73) QSurface::OpenGLSurface exposed false
```
When I do the duplication
```
Debug [Main]: WWaveformViewer(0x1f7559809f0) resizeEvent
Debug [Main]: WWaveformViewer(0x1f7559809f0) isVisible() true QSize(1507, 73)
Debug [Main]: WWaveformViewer(0x1f7559809f0) isVisible() WWidgetGroup(0x1f75f0f3200, name="Waveform") QSize(1585, 73)
Debug [Main]: WWaveformViewer(0x1f7559809f0) widget isVisible() true QSize(22, 10)
Debug [Main]: WWaveformViewer(0x1f7559809f0) widget isVisible() true QSize(1507, 73)
Debug [Main]: QWidgetWindow(0x1f76a601070, name="QWidgetClassWindow") vis true QSize(1507, 73) QSurface::OpenGLSurface exposed true
```

Note: I get 8 resizeEvent with either of both cases. 
Without duplication, all 8 show exposed false.
When I duplicate, the three last show exposed true (which means the exposing happens after the first resizing once the widged is visible), but does not happen without the duplication.

I can share the logs but.. where could I put them?

daschuer replied:
e.g. https://pastebin.com/

daschuer reviewed (commented):
OK, everything is still working here. I have left some hints in inline. 
Nothing blocking.

JosepMaJAZ committed:
Merge remote-tracking branch 'remotes/upstream/master' into GL_Qt514_windows-fix

JosepMaJAZ committed:
Removing the "call" from the sets, as suggested in the PR.

JosepMaJAZ committed:
Removed makeCurrent() from QGLWidget destructors:
Other gl waveforms don't have it, and it crashed on Mixxx close when I tested glsimplewaveformwidget because currentContext() (or was it context()?) was null
Also added makeCurrent() to qthsvwaveformwidget constructor, which was missing.

JosepMaJAZ committed:
Added ifdefs for QT version and OS as suggested in the PR, plus some other suggestions.

Be-ing commented on src/waveform/waveformwidgetfactory.h:
```diff
@@ -83,10 +84,21 @@ class WaveformWidgetFactory : public QObject, public Singleton<WaveformWidgetFac
+    /// Returns false and sets EmtpyWaveform if type is invalid
     bool setWidgetType(WaveformWidgetType::Type type);
-    bool setWidgetTypeFromHandle(int handleIndex);
+    /// Changes the widged type to that loaded from config and recreates them.
```
```suggestion
    /// Changes the widget type to that loaded from config and recreates them.
```

Be-ing commented:
@daschuer ready for merge?

Be-ing commented:
I double checked this on Linux and did not see a regression.

JosepMaJAZ committed:
fix comment

Co-Authored-By: Be <be.0@gmx.com>

daschuer reviewed (approved):
I have tried to wrap my head around the log and the issue to find and fix the root cause.
But it is too hard to do this without the effected system. 
Since this has no harm here as well lets just merge it and proceed to the release.

daschuer commented:
LGTM 
@Holzhaus: Are your review comments issued as well?

Be-ing commented:
ping @Holzhaus, ready for merge?
